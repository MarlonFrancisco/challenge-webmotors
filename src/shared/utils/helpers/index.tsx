import React from "react";
import { ThemeProvider } from "styled-components";
import { theme } from "../../styled";

type WithThemeProps = {};

export const WithTheme: React.FC<WithThemeProps> = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);
