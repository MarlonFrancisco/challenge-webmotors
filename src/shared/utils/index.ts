export type Options = {
  withoutPrefix: boolean;
  splitted: boolean;
};

export const formatCurrency = (
  value = 0,
  options?: Options
): string | string[] => {
  const formatted = new Intl.NumberFormat("pt-BR", {
    style: "currency",
    currency: "BRL",
  }).format(value);
  let apllyOptions: string | string[] = formatted;

  if (!options) {
    return formatted;
  }

  if (options.withoutPrefix) {
    apllyOptions = apllyOptions.replace("R$", "").trim();
  }

  if (options.splitted) {
    apllyOptions = apllyOptions.split(",");
  }

  return apllyOptions;
};
