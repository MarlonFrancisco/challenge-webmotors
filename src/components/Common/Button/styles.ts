import styled from "styled-components";
import { Theme } from "../../../shared/@types";

export const SCButton = styled.button`
  border: none;
  border-radius: 4px;
  padding: 8px 20px;

  ${(props: {
    variant: string;
    theme: Theme;
    color: "primary" | "secondary";
  }) =>
    props.variant === "outlined" &&
    `
        background: none;
        border: 2px solid ${props.theme.palette[props.color].main};
        color: ${props.theme.palette[props.color].main};
    `}

  ${(props: {
    variant: string;
    theme: Theme;
    color: "primary" | "secondary";
  }) =>
    props.variant === "contained" &&
    `
            background: ${props.theme.palette[props.color].main};
            border: 2px solid ${props.theme.palette[props.color].main};
            color: #fff;
        `}

    ${(props: {
    variant: string;
    theme: Theme;
    color: "primary" | "secondary";
  }) =>
    props.variant === "text" &&
    `
                    background: transparent;
                    border: 2px solid transparent;
                    color: ${props.theme.palette.text.primary};
                `}
`;
