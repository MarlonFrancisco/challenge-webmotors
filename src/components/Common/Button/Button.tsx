import React, { CSSProperties } from "react";
import { Loader } from "../Loader";
import { SCButton } from "./styles";

type ButtonProps = {
  variant?: "contained" | "outlined" | "text";
  color?: "primary" | "secondary";
  style?: CSSProperties;
  onClick?: () => void;
  fetching?: boolean;
};

const Button: React.FC<ButtonProps> = ({
  children,
  variant = "contained",
  color = "primary",
  fetching,
  ...rest
}) => {
  return (
    <SCButton variant={variant} color={color} {...rest}>
      {fetching ? <Loader /> : children}
    </SCButton>
  );
};

export default Button;
