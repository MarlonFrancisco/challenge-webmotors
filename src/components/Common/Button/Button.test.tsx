import { render } from "@testing-library/react";
import React from "react";
import { WithTheme } from "../../../shared/utils/helpers";
import Button, { ButtonProps } from "./Button";

const Wrapper = (props: ButtonProps) => (
  <WithTheme>
    <Button {...props} />
  </WithTheme>
);

describe("Button", () => {
  let props: ButtonProps & { children: React.ReactNode };

  beforeEach(() => {
    props = {
      fetching: false,
      variant: "contained",
      color: "primary",
      children: <p>ButtonComponent</p>,
    };
  });

  it("should render without an error", () => {
    const { container } = render(<Wrapper {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should render Loader component on fetching props is true", () => {
    const { queryByTestId } = render(<Wrapper {...props} fetching />);
    expect(queryByTestId("loader-test")).not.toBeNull();
  });

  it("shouild render children component", () => {
    const { queryByText } = render(<Wrapper {...props} />);

    expect(queryByText("ButtonComponent").innerHTML).not.toBeNull();
  });
});
