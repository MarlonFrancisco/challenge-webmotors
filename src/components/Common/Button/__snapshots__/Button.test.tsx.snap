// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render without an error 1`] = `
<div>
  <button
    class="sc-gtssRu bSbDmz"
    color="primary"
  >
    <p>
      ButtonComponent
    </p>
  </button>
</div>
`;
